Sources/MapboxNavigation/NavigationViewController.swift

```diff
@@ -314,49 +367,47 @@ open class NavigationViewController: UIViewController, NavigationStatusPresenter
         navigationService.stop()
     }
     
+    fileprivate func handleCancelAction() {
+        if delegate?.navigationViewControllerDidDismiss(self, byCanceling: true) != nil {
+            // The receiver should handle dismissal of the NavigationViewController
+        } else {
+            dismiss(animated: true, completion: nil)
+        }
+    }
+    
     // MARK: - Setting-up methods
     
     func setupStyleManager(_ navigationOptions: NavigationOptions?) {
         styleManager = StyleManager()
         styleManager.delegate = self
         styleManager.styles = navigationOptions?.styles ?? [DayStyle(), NightStyle()]
         
-        // Manually update the map style since the `RouteMapViewController` missed the
-        // "map style change" notification when the style manager was set up.
         if let currentStyle = styleManager.currentStyle {
             updateMapStyle(currentStyle)
         }
     }
     
-    func addRouteMapViewController(_ navigationOptions: NavigationOptions?) {
-        let mapViewController = RouteMapViewController(navigationService: self.navigationService,
-                                                       delegate: self,
-                                                       topBanner: addTopBanner(navigationOptions),
-                                                       bottomBanner: addBottomBanner(navigationOptions))
-        mapViewController.destination = route.legs.last?.destination
-        mapViewController.view.pinInSuperview()
-        mapViewController.reportButton.isHidden = !showsReportFeedback
-        mapViewController.view.translatesAutoresizingMaskIntoConstraints = false
-        
-        self.mapViewController = mapViewController
+    func setupControllers(_ navigationOptions: NavigationOptions?) {
+        arrivalController = ArrivalController()
+        navigationViewHelper = NavigationViewHelper()
+        cameraController = CameraController()
+        overlayController = OverlayController()
         
-        embed(mapViewController, in: view) { (parent, map) -> [NSLayoutConstraint] in
-            return map.view.constraintsForPinning(to: parent.view)
-        }
+        arrivalController?.navigationViewData = self
+        navigationViewHelper?.navigationViewData = self
+        cameraController?.navigationViewData = self
+        overlayController?.navigationViewData = self
```

Comment:
I think we should minimize usage of backpointers if possible: classes that contain backpointers either have to account for the possibility of them going away at any time, mucking up the code, or use implicitly unwrapped optionals, which can be risky. A delegate protocol – or in this case, a data source protocol – would similarly require optional method calls, but at least the communication between the delegator and delegate is very explicit and isn’t as sensitive to object lifecycles, which makes further refactoring more straightforward.

Reply:
This is the best way of communication I ended up with. What do you think could work better?